### PR TITLE
feat: thing manifest schema + metadata assembler

### PR DESCRIPTION
**Pull request by devonjones**
## Summary

Ticket `openforge_catalog-d7a` — the "what goes in a thing and what it says" layer of the Thingiverse epic (`4kx`). Unblocks the sync engine (`ad5`) together with #225.

- **`openforge/thingiverse/manifest.py`** — per-thing YAML manifests, strictly validated (unknown keys/sections/select-keys all fail fast — typos die at load, not at publish). Model entries:
  - `select:` tag queries — the killer feature: newly scanned files matching the query flow into the thing on the next sync run, zero manual entry. Semantics match the catalog's tag engine: `require` = exact-match ALL, `accept` = hierarchical-prefix ANY, `deny` = exact exclude.
  - explicit `md5:` / `full_name:` refs (ambiguous full_names rejected with a pointer to use md5)
  - `images`/`zips`/`others` as local paths resolved relative to the manifest
- **`openforge/thingiverse/templates/openforge2.yaml`** — product-line defaults (license, category, base tags, description boilerplate) ported from the old v1 tooling's `new.openforge2.json`. Boilerplate lives once in the template, so changing the copy re-syncs every managed thing's description — the old `tv_update_description` bulk-edit workflow, automated. ⚠️ **The boilerplate copy is the original ~2016 text (reddit/twitter links) — needs Devon's review before first production publish.**
- **`openforge/thingiverse/assembler.py`** — `assemble_thing(curs, manifest)` → API-neutral payload: merged metadata (template ← manifest overrides), resolved blueprint rows (deduped by md5), verified local files, and `metadata_hash` (canonical sha256 that feeds `thingiverse_things.remote_metadata_hash` for drift detection). Fail-fast everywhere: unknown template, selector matching nothing, selector hitting the truncation ceiling, missing refs, missing files.

One subtle bug caught during dev worth flagging: `tag_search_blueprints` takes `{"tag": ...}` dicts — bare strings **silently no-op** the filters (`"tag" in "texture|cave"` is just a substring check). The assembler wraps correctly and carries a comment.

## Test plan

- [x] 42 new tests: 13 manifest validation, 29 assembler (template load, override/merge/compose semantics, DB-backed selector require/accept/deny behavior, md5/full_name resolution, dedupe, local-file resolution, metadata_hash stability)
- [x] `ruff check .` clean; full suite 402 passed (same 4 pre-existing `test_incremental` local failures; CI green on `test`)

Beads: `openforge_catalog-d7a` (closed via `bd close` on merge).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: openforge/thingiverse/manifest.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    _validate_top_level(raw, path)
+    manifest = {
+        "name": raw["name"],
+        "template": raw.get("template", DEFAULT_TEMPLATE),
+        "description": raw.get("description", ""),
+        "category": raw.get("category"),
+        "license": raw.get("license"),
+        "tags": _validate_string_list(raw.get("tags", []), "tags", path),
+        "files": _validate_files(raw.get("files", {}), path),
+        "manifest_path": str(path),
```
Comment (devonjones) [inline openforge/thingiverse/manifest.py:101]: 🤖 **Claude Code** (dead-code-reviewer):


[dead-code-reviewer] [P3] `manifest_path` is set on the returned manifest dict but has no consumer anywhere in the repo — assembler.py only reads `manifest_dir` (line 201), never `manifest_path`; neither test file (`tests/test_thingiverse_manifest.py`, `tests/test_thingiverse_assembler.py`) asserts on it. grep evidence: `grep -rn manifest_path --include='*.py' .` returns only the two lines in this file (definition comment on L75, assignment on L99) plus a docstring mention — no reads. This may be intentional groundwork for the sync engine (ad5) or CLI (lpe), which will need to know which manifest a thing came from, so I'm not asking for removal — just flagging that it's currently 100% write-only in case it was meant to be threaded through and wasn't, or should get a code comment noting the future consumer like `manifest_dir`'s docstring does.

File: openforge/thingiverse/manifest.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    Raises:
+        ManifestError: On missing file, bad YAML, or validation failure
+    """
+    path = Path(path)
+    try:
+        raw = safe_load(path.read_text())
+    except FileNotFoundError as e:
+        raise ManifestError(f"manifest not found: {path}") from e
+    except Exception as e:
```
Comment (devonjones) [inline openforge/thingiverse/manifest.py]: 🤖 **Claude Code** (error-handling-reviewer):


[error-handling-reviewer] [P3] The `except Exception as e:` on this line catches everything `path.read_text()` and `safe_load()` can raise — not just YAML syntax errors. `read_text()` alone can also raise `PermissionError`, `IsADirectoryError`, or `UnicodeDecodeError` (e.g. a manifest with non-UTF-8 bytes), all of which get relabeled 'is not valid YAML' even though the actual problem is unrelated to YAML parsing. Since this is a single-Lambda deployment where CloudWatch tracebacks are the only forensics, a mislabeled error class costs real debugging time. The exception is still chained (`from e`) so it's not a P1/P2 swallow, but consider narrowing to `except yaml.YAMLError as e:` (optionally + `UnicodeDecodeError`) so genuinely unexpected failures (permissions, wrong file type) surface with their real type/traceback instead of being folded into a YAML-syntax message.

File: openforge/thingiverse/assembler.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+            resolved.append(_resolve_full_name(curs, entry["full_name"]))
+    return _dedupe_models(resolved)
+
+
+def _resolve_select(curs: cursor, select: Dict) -> List[Dict]:
+    # tag_search_blueprints takes {"tag": ...} dicts (the blueprint-config
+    # parts shape), not bare strings — bare strings silently no-op.
+    rows = tag_sql.tag_search_blueprints(
+        curs,
+        accept=[{"tag": t} for t in select["accept"]],
```
Comment (devonjones) [inline openforge/thingiverse/assembler.py]: 🤖 **Claude Code** (test-coverage-reviewer):


[test-coverage-reviewer] [P1] `_resolve_select`'s `accept` key (prefix-ANY match, per the manifest.py docstring) is never exercised end-to-end. `TestModelResolution` in tests/test_thingiverse_assembler.py covers `require` (exact match) and `deny`, but no test builds a manifest with `select.accept` and verifies a model whose tag is a *child* of the accepted prefix (e.g. accept: ['shape|floor'] should match a blueprint tagged 'shape|floor|corner') gets selected, nor that require's exact-match/accept's prefix-match difference actually holds through tag_search_blueprints. This is new wiring (building `[{"tag": t}, ...]` for accept) with real semantic behavior, not just glue — add a test analogous to test_select_by_tags but using accept with a hierarchical child tag.

File: openforge/thingiverse/assembler.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    if not rows:
+        raise AssemblyError(f"no blueprint with md5 {md5}")
+    return rows[0]
+
+
+def _resolve_full_name(curs: cursor, full_name: str) -> Dict:
+    rows = blueprint_sql.get_blueprints_by_full_name(curs, full_name)
+    if not rows:
+        raise AssemblyError(f"no blueprint with full_name {full_name!r}")
+    if len(rows) > 1:
```
Comment (devonjones) [inline openforge/thingiverse/assembler.py:211]: 🤖 **Claude Code** (test-coverage-reviewer):


[test-coverage-reviewer] [P1] The ambiguous-full_name branch (`len(rows) > 1`) in `_resolve_full_name` is uncovered (coverage report: line 180 missed). `get_blueprints_by_full_name` has no DB uniqueness constraint on full_name, so this is reachable, real business logic in this PR's new code, and cheap to test: call `insert_model` twice with the same `full_name` and different md5s, then assert `AssemblyError` with match='is ambiguous'. tests/test_thingiverse_assembler.py's TestModelResolution currently only covers the not-found and single-match paths for full_name (test_explicit_full_name, test_missing_full_name_raises).

File: openforge/thingiverse/assembler.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        accept=[{"tag": t} for t in select["accept"]],
+        require=[{"tag": t} for t in select["require"]],
+        deny=[{"tag": t} for t in select["deny"]],
+        limit=SELECT_LIMIT,
+        models=True,
+        blueprints=False,
+    )
+    if not rows:
+        raise AssemblyError(f"selector matched no models: {select}")
+    if len(rows) >= SELECT_LIMIT:
```
Comment (devonjones) [inline openforge/thingiverse/assembler.py:193]: 🤖 **Claude Code** (test-coverage-reviewer):


[test-coverage-reviewer] [P1] SELECT_LIMIT ceiling guard (`len(rows) >= SELECT_LIMIT`, line 161) is uncovered — the docstring/comment frames this as an explicit anti-silent-truncation safeguard, which is exactly the kind of boundary a future off-by-one edit (>= vs >) would silently break without a test. It doesn't need 10,000 real rows: `monkeypatch.setattr(openforge.thingiverse.assembler, "SELECT_LIMIT", 2)` then insert 2 matching blueprints and assert `AssemblyError` with match='ceiling'. Currently untested in tests/test_thingiverse_assembler.py::TestModelResolution.

File: openforge/thingiverse/manifest.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+
+def _validate_string_list(value, field: str, path: Path) -> List[str]:
+    if not isinstance(value, list) or not all(isinstance(v, str) for v in value):
+        raise ManifestError(f"manifest {path}: '{field}' must be a list of strings")
+    return value
+
+
+def _validate_files(files, path: Path) -> Dict:
+    if not isinstance(files, dict):
```
Comment (devonjones) [inline openforge/thingiverse/manifest.py:126]: 🤖 **Claude Code** (test-coverage-reviewer):


[test-coverage-reviewer] [P3] Coverage report shows lines 125, 147, 161, 187 uncovered — the isinstance-guard raise in `_validate_files` (files not a mapping), `_validate_model_entry` (entry not a mapping), `_validate_select` (select not a mapping), and `_validate_path_entry` (path not a non-empty string). These mirror the pattern already tested at the top level (test_non_mapping_raises) but aren't covered at their own nesting depth. Low materiality (trivial one-line type guards, same shape as an already-tested sibling check) — optional parametrize additions, not blocking.

File: openforge/thingiverse/assembler.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+
+def load_template(name: str) -> Dict:
+    """Load a named metadata template bundled with the package.
+
+    Raises:
+        AssemblyError: If no template with that name ships in
+            openforge/thingiverse/templates/
+    """
+    resource = impresources.files("openforge.thingiverse.templates").joinpath(
```
Comment (devonjones) [inline openforge/thingiverse/assembler.py:43]: 🤖 **Claude Code** (importlib-resources-reviewer):


[importlib-resources-reviewer] [P3] Verified: this matches the impresources convention from openforge/db/fixtures/__init__.py (no Path(__file__) navigation). Confirmed working: `python3 -c "from importlib import resources as r; print(r.files('openforge.thingiverse.templates').joinpath('openforge2.yaml').read_text()[:50])"` succeeds — templates/ does NOT need an __init__.py; it resolves as an implicit namespace package under the openforge.thingiverse regular package.

One gap worth a look: setup.py's package_data only lists `openforge/db/fixtures` and `openforge/openapi` (and `packages=["openforge"]` doesn't even list subpackages via find_packages()). The new openforge/thingiverse/templates/openforge2.yaml isn't declared anywhere in setup.py. It ships fine today only because Dockerfile.api.deploy does `COPY openforge ./openforge` (raw directory copy, no .dockerignore) before `python setup.py install", so the actual runtime import path is the raw source tree, not whatever setup.py's package_data would produce. If the deploy path ever changes to installing from a built wheel/sdist instead of copying raw source, this file (and the rest of openforge/thingiverse, which was never added to setup.py's packages list either) would silently go missing in prod. Not a blocker for this PR since it's consistent with the pre-existing (already stale) packaging config, but flagging since it's the exact failure mode this review is meant to catch.

**Review by devonjones — COMMENTED** [review]
## Review scope

Domain correctness of the manifest → assembled thing payload for the Thingiverse epic (`openforge_catalog-d7a`, unblocking `ad5`), selector semantics vs. `openforge/db/sql/tags.py`, `metadata_hash` stability, and house style. Complexity/error-handling/test-coverage/dead-code/importlib-resources/credentials are covered by other passes and skipped here. Verified several claims experimentally against a live test DB (`pytest` + `test_db` fixture), not just by reading.

## Critical

1. **`select.accept` with multiple entries is ANDed, not ORed — contradicts the manifest.py docstring's documented "ANY" semantics.**

   The docstring says: *"`accept`: model must carry at least one tag in the listed subtree, matched by hierarchical prefix ANY."* That's true for a **single** accept entry (verified: `accept: ['shape|floor']` correctly matches a blueprint tagged `shape|floor|corner` via prefix match). But with **multiple** entries it's a different story.

   `_query_tags_include` (`openforge/db/sql/tags.py`) adds one `JOIN tags AS tags_N` per accept entry and just concatenates all the WHERE conditions — there's no OR grouping across entries, only within a single entry's own position checks. That means a blueprint must satisfy **every** listed accept subtree simultaneously (same AND behavior as `require`), not any one of them.

   I confirmed this by inserting two blueprints, each carrying exactly one of two sibling texture tags, and running:
   ```yaml
   select:
     accept: ['texture|cave', 'texture|dungeon_stone']
   ```
   Expected (per docstring): both blueprints match (OR). Actual: `AssemblyError: selector matched no models` — because no single blueprint carries *both* texture subtrees at once.

   This means the most natural use of a multi-value `accept` list — "any of these sibling subtrees" (e.g. several textures, several sizes) — can never match anything unless a blueprint happens to be tagged under every listed subtree at once. Since the assembler fails fast on empty selector results, this surfaces as an immediate hard error rather than silent corruption, but it means an entire class of manifest expression the docstring promises doesn't work. There is no test anywhere (`test_thingiverse_assembler.py`, `test_thingiverse_manifest.py`) exercising `accept` with more than one entry, so this shipped unnoticed.

   This needs a decision before `ad5`/`lpe` manifests start relying on it: either `_query_tags_include`/`_query_tags_basics` need real OR-across-entries support for `accept` (touches shared tag-search code used elsewhere by the app, e.g. the search API), or the assembler needs to run one sub-select per accept entry and union the results itself, or (minimal fix) the docstring should be corrected to state accept is ALL-across-entries like require (in which case `accept` differs from `require` only by prefix-vs-exact matching, which is a much narrower and less useful primitive than advertised).

## Medium

2. **Model row shape differs depending on resolution path.** `_resolve_select` gets rows via `tag_search_blueprints`'s `SELECT *` (all `blueprints` columns, including `search_text`, added in schema v4). `_resolve_md5`/`_resolve_full_name` get rows via `get_blueprints_by_md5`/`get_blueprints_by_full_name`'s explicit column list (no `search_text`). Both paths run rows through `_convert_config`, so the fields the sync engine actually needs (`file_md5`, `file_name`, `full_name`, `storage_address`) are present and consistent either way — but a manifest mixing `select:` entries with `md5:`/`full_name:` entries for the same thing produces a `files.models` list whose dicts don't all have the same keys. Low risk today since nothing downstream inspects `search_text`, but worth normalizing to one explicit projection so `files.models` is shape-stable regardless of how each entry was resolved.

## Minor

3. **`_resolve_md5`/`_resolve_full_name` don't filter by `blueprint_type='model'`** the way `_resolve_select` explicitly does (`models=True, blueprints=False`, with a comment explaining why). I traced the fixtures/scanner pipeline (`openforge/db/fixtures/utils.py:munge_blueprint`, `openforge/db/fixtures/incremental.py`) and confirmed composite `blueprint_type='blueprint'` rows never populate `file_md5`/`full_name` today (they're config-only compositions, keyed by `blueprint_name` instead), so this is currently unreachable — a real md5/full_name literal can never equal a `NULL` column. But nothing enforces that invariant at the schema level (no CHECK constraint), and the admin blueprint API (`openforge/app/routes/blueprints.py`, `openforge/openapi/schemas/blueprint.yaml`) would accept a row with `blueprint_type='blueprint'` and `full_name`/`file_md5` both set if anyone ever created one by hand. Adding the same type filter to `_resolve_md5`/`_resolve_full_name` is a one-line-each defense-in-depth fix.

4. **`_dedupe_models` keys on `row.get("file_md5")`**, which trusts every resolved row has a non-null, unique md5. If that were ever violated (only reachable via #3), two distinct rows with `file_md5 = None` would silently collapse into one instead of erroring. Moot once #3 is closed, but the function has no explicit assertion of the invariant it relies on.

5. **`manifest["license"] or template.get("license")` (and same pattern for `category`)** treats an explicit-but-falsy manifest override (e.g. `license: ""`) the same as "not set," silently falling back to the template value. Edge case, probably never hit since these fields are either meaningfully set or omitted (`None`) in practice, but it's a subtle `""` vs `None` trap worth a one-line comment for future maintainers.

## Positive

- `storage_address`, `file_md5`, and `file_name` are present on every resolved model row (verified for both selector and explicit-ref paths), so `ad5` has everything it needs to know where to fetch a model's bytes from (R2 via `storage_address`) and what to name it on upload (`file_name`) without extra lookups.
- The `{"tag": t}` wrap for `tag_search_blueprints` is correct and well-commented. Verified against `tags.py`: bare strings really would silently no-op, since `_query_tags_include`/`_query_tags_deny` gate on `if "tag" in req`/`if "tag" in d`, and a bare string like `"texture|cave"` makes `"tag" in "texture|cave"` a truthy substring check that then crashes/misbehaves rather than doing a real filter. Good catch, correctly guarded against.
- `require` (exact ALL across entries) and `deny` (exact match, excluded if any listed tag present) semantics both match the docstring exactly — verified experimentally and by reading `_query_tag_require`/`_query_tags_deny`.
- `metadata_hash` is properly decoupled from file selection — only the `metadata` sub-dict is hashed, not `files`, so selector/file drift isn't conflated with "should this thing's metadata re-push," matching `ad5`'s separate file-vs-metadata diffing design. `sort_keys=True` handles dict-key ordering, and tag-list order is deterministic (template-then-manifest, stable case-insensitive dedupe), so a template boilerplate or tag change reliably changes the hash as intended.
- Fail-fast validation in `manifest.py` (unknown top-level keys, unknown file sections, unknown select keys, ambiguous `full_name`) is thorough and in keeping with the project's "no silent failures" philosophy.
- `SELECT_LIMIT` truncation-ceiling check (`len(rows) >= SELECT_LIMIT`) is correctly reasoned given `tag_search_blueprints`'s LIMIT-based pagination and the catalog's known size (~1,400 designs).
- `ruff check`/`ruff format --check` both clean; docstrings, private-helper decomposition, and function sizes match `openforge/CLAUDE.md`'s style guide.

## Recommendations

1. Get a ruling from Devon on `accept`'s intended algebra (ANY vs. ALL across list entries) and fix either the SQL or the docstring to match — this is the one item that actually blocks safe consumption of manifests by anyone writing selectors for `ad5`/`lpe`.
2. Add a regression test for multi-entry `accept` (none exists today) once semantics are settled — this gap is squarely why the mismatch shipped.
3. Normalize model row shape between `_resolve_select` and `_resolve_md5`/`_resolve_full_name` to one explicit column projection.
4. Add `blueprint_type='model'` filtering to `_resolve_md5`/`_resolve_full_name` for defense-in-depth.

## Overall

The plumbing that `ad5` actually depends on — `storage_address`/`file_md5`/`file_name` flow-through, `metadata_hash` design, fail-fast validation — is solid and gives the sync engine what it needs. The one real correctness issue is the `accept` selector's actual AND semantics contradicting its documented ANY semantics; that's worth resolving before other manifests start writing multi-value `accept` lists expecting OR behavior. Everything else here is minor hardening, not a blocker.

File: openforge/thingiverse/manifest.py (hunk 1 above)
Comment (devonjones) [reply openforge/thingiverse/manifest.py:101]: Won't fix - manifest_path is deliberate groundwork: thingiverse_things.manifest_path (schema v17, PR #225) is its consumer — the sync engine (ad5) records which manifest a managed thing came from.

File: openforge/thingiverse/manifest.py (hunk 2 above)
Comment (devonjones) [reply openforge/thingiverse/manifest.py]: Fixed - narrowed to except (YAMLError, UnicodeDecodeError); PermissionError/IsADirectoryError now propagate with their real type instead of being mislabeled as YAML problems. Comment documents the deliberate narrowing.

File: openforge/thingiverse/assembler.py (hunk 3 above)
Comment (devonjones) [reply openforge/thingiverse/assembler.py]: Fixed - and it caught a real bug (the Claude reviewer independently confirmed multi-accept was ANDed by the engine, contradicting the docs). _resolve_select now runs one query per accept entry and unions, giving the documented ANY semantics; test_multiple_accepts_are_any_not_all pins it, and test_accept_matches_by_hierarchical_prefix covers single-accept prefix matching.

File: openforge/thingiverse/assembler.py (hunk 4 above)
Comment (devonjones) [reply openforge/thingiverse/assembler.py:211]: Fixed - test_ambiguous_full_name_raises inserts two models sharing a full_name and asserts the 'ambiguous' AssemblyError.

File: openforge/thingiverse/assembler.py (hunk 5 above)
Comment (devonjones) [reply openforge/thingiverse/assembler.py:193]: Fixed - test_select_limit_ceiling_raises monkeypatches SELECT_LIMIT to 2 and asserts the ceiling error fires (>= boundary exercised).

File: openforge/thingiverse/manifest.py (hunk 6 above)
Comment (devonjones) [reply openforge/thingiverse/manifest.py:126]: Won't fix - per your own P3 framing: the four isinstance guards mirror the already-tested sibling pattern (test_non_mapping_raises) and are defensive-only; per-guard tests add maintenance weight without meaningful signal.

File: openforge/thingiverse/assembler.py (hunk 7 above)
Comment (devonjones) [reply openforge/thingiverse/assembler.py:43]: Out of scope - tracked in openforge_catalog-li2. Pre-existing packaging gap (whole openforge.thingiverse subpackage undeclared in setup.py; ships today via the Dockerfile raw copy). Ticket carries the find_packages()+package_data fix and the wheel/sdist failure mode.

File: openforge/thingiverse/assembler.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+        raise AssemblyError(
+            f"{ref} resolves to a {row.get('blueprint_type')!r} blueprint, not a model"
+        )
+    return row
+
+
+def _project_model(row: Dict) -> Dict:
+    """Project a blueprint row to the stable payload shape."""
+    if not row.get("file_md5"):
+        raise AssemblyError(
```
Comment (devonjones) [inline openforge/thingiverse/assembler.py:231]: 🤖 **Claude Code** (test-coverage-reviewer):


[P2] New in df77937: `_project_model`'s fail-fast branch (raise when `row.get("file_md5")` is falsy — line 231) is uncovered per this run's coverage report (assembler.py 99%, only line 231 missing). This isn't a mirrored defensive guard like the won't-fixed manifest.py isinstance checks — it's new business logic the commit message calls out explicitly ('fail-fast on models missing file_md5'), and it's reachable: `insert_blueprint`'s default for `file_md5` is SQL NULL (openforge/db/sql/blueprints.py:14), and the unique constraint on `file_md5` doesn't block multiple NULLs, so a 'model'-typed row with no file_md5 is realistic (e.g. mid-sync or bad data) and would reach this path via both `select` and explicit `full_name`/`md5` refs. Cheap to test analogous to test_explicit_ref_to_non_model_raises: insert_model(..., md5=None or '') via create_test_blueprint, reference it by full_name or select, assert AssemblyError matching 'no file_md5'.

Comment (devonjones) [reply openforge/thingiverse/assembler.py:231]: Fixed - test_model_without_md5_raises inserts a model with file_md5 forced to NULL (the test helper generates one otherwise) and asserts the fail-fast AssemblyError. 36 tests passing, assembler.py at 100%.
